src/ore/src/cgroup.rs

```diff
@@ -0,0 +1,251 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Linux cgroup detection utilities.
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+
+/// An entry in /proc/self/cgroup
+#[derive(Debug, PartialEq)]
+pub struct CgroupEntry {
+    subsystems: Vec<String>,
+    root: PathBuf,
+}
+
+impl CgroupEntry {
+    fn from_line(line: String) -> Option<CgroupEntry> {
+        let mut fields = line.split(':');
+        let subsystems = fields
+            .nth(1)?
+            .split(',')
+            .filter(|s| !s.is_empty())
+            .map(|s| s.to_owned())
+            .collect();
+        let root = PathBuf::from(fields.next()?);
+        Some(CgroupEntry { subsystems, root })
+    }
+}
+
+/// Parses /proc/self/cgroup into a Vec<CgroupEntry>, if the file exists.
+pub fn parse_proc_self_cgroup() -> Option<Vec<CgroupEntry>> {
+    let file = File::open("/proc/self/cgroup").ok()?;
+    let file = BufReader::new(file);
+    Some(
+        file.lines()
+            .filter_map(|result| result.ok())
+            .filter_map(CgroupEntry::from_line)
+            .collect(),
+    )
+}
+
+/// An entry in /proc/self/mountinfo
+#[derive(Debug, PartialEq)]
+pub struct Mountinfo {
+    root: PathBuf,
+    mount_point: PathBuf,
+    fs_type: String,
+    super_opts: Vec<String>,
+}
+
+impl Mountinfo {
+    fn from_line(line: String) -> Option<Mountinfo> {
+        // https://www.kernel.org/doc/Documentation/filesystems/proc.txt
+        let mut split = line.split(" - ");
+
+        let mut mount_fields = split.next()?.split(' ');
+        let root = PathBuf::from(mount_fields.nth(3)?);
+        let mount_point = PathBuf::from(mount_fields.next()?);
+
+        let mut fs_fields = split.next()?.split(' ');
+
+        let fs_type = fs_fields.next()?.split('.').next()?.to_owned();
+        let super_opts: Vec<String> = fs_fields
+            .nth(1)?
+            .split(',')
+            .filter(|s| !s.is_empty())
+            .map(|s| s.to_owned())
+            .collect();
+
+        Some(Mountinfo {
+            root,
+            mount_point,
+            fs_type,
+            super_opts,
+        })
+    }
+}
+
+/// Parses /proc/self/mountinfo into vectors of Mountinfo objects,
+/// returning (v2_mounts, v1_mounts).
+pub fn parse_proc_self_mountinfo() -> Option<(Vec<Mountinfo>, Vec<Mountinfo>)> {
+    let file = File::open("/proc/self/mountinfo").ok()?;
+    let file = BufReader::new(file);
+    Some(
+        file.lines()
+            .filter_map(|result| result.ok())
+            .filter_map(Mountinfo::from_line)
+            .filter(|mi| mi.fs_type == "cgroup" || mi.fs_type == "cgroup2")
+            .partition(|mi| mi.fs_type == "cgroup2"),
+    )
+}
+
+/// Represents a cgroup memory limit, with both ram and swap maximums if they exist.
```

Comment:
What happens if a process is not explicitly assigned a cgroup? Based on the changes in main it seems like this will be empty in that case? Could you document how it behaves?

Reply:
Yes, the fields will be `None` if the limit does not exist, or if the platform does not support cgroups. I'll add that to the comment.